features/370-auth-ui/design.md

```diff
@@ -0,0 +1,529 @@
+      * [Example of valid Token Request](#example-of-valid-token-request)
+   * [Token Response](#token-response)
+      * [id_token](#id_token)
+      * [token_type](#token_type)
+      * [access_token](#access_token)
+      * [expires_in](#expires_in)
+      * [refresh_token](#refresh_token)
+      * [scope](#scope-1)
+      * [Example of Token Response](#example-of-token-response)
+* [Security considerations](#security-considerations)
+   * [Being embedded in an HTML iframe](#being-embedded-in-an-html-iframe)
+* [Appendix](#appendix)
+
+## Background
+
+Auth UI is a standalone component that aims to be a OpenID Provider. It is a traditional web application that does **NOT** share the session with Auth Gear. However, it provides access to most of the Auth Gear features.
+
+## Relationship among Auth UI, Auth Gear, API client, Microservices and Data
+
+### Past Relationship
+
+In the past, there is no Auth UI, Auth Gear and microservices access Data via the API key of a API client.
+
+### Diagram of Past Relationship
+
+Below is a diagram to illustrate to the past relationship.
+
+```
+                 +--------+
+                 |  Data  |
+                 +----+---+
+                      ^
+                      |
+                      | Access
+                      |
+               +------+------+
+               |  Auth Gear  |
+               +------+------+
+                      |
+                      | Validate API key
+                      |
+              +-------+------+
+              |  API Client  |
+              +-------+------+
+                      ^
+                      |
+         +------------+----------+  Identify with API Key
+         |                       |
+         |                       |
+         |                       |
++--------+--------+          +---+---+
+|  Microservices  |          |  SDK  |
++-----------------+          +-------+
+```
+
+### New Relationship
+
+After the introduction of Auth UI, it acts an OpenID Provider (OP). API Client is now treated as Relying Party (RP). Microservices and Auth Gear use the API Key of a API Client to identify themselves as RP.
+
+Auth UI does not support Authorization at the moment. To be backward compatible, existing access tokens can access Data without any restriction. That means the user can change password, list sessions of **ALL** API clients, etc, via Auth Gear. The new access tokens issued by Auth UI are intended to be consumed in the same fashion so the new and the old behave the same.
+
+In the future Auth UI can act as an OP for another Skygear App. At that moment Authorization must be introduced. Existing API clients are considered as first party and have the largest scopes. The RP Skygear App must register itself as a new kind of API client with the OP Skygear App. This is analogous to using access token issued by Google cannot access Google API to change Google account password unless the user authorize the corresponding scope.
```

Comment:
`In the future Auth UI can act as an OP for another Skygear App`

Thought experiment — Can 3rd parties Skygear App support “the Facebook Messenger” flow?

It works something like this (assume Skygear-App2 using Skygear-App1 auth):
* Visit www.skygear-app2.com, it display “Ben Cheng From skygear-app1? Click here to Login”
* Redirect to AuthUI of skygear-app1, login directly without input of password

Reply:
[The OpenID Connect spec has something similar but not exactly the same](https://openid.net/specs/openid-connect-session-1_0.html#OPiframe).

Basically iframes and postMessage are used to detect session status in the OP. This technique enables your Point 1.

Point 2 can be achieved with `prompt=none`. (prompt=none is mandatory anyway If OP supports the above iframe)

Reply:
icic, that meant if we do it in future it will be a proprietary implementation instead of a standard one... maybe worth to open a feature issue about it too? I don't think it is a must for now. More like good to have...

Reply:
Created https://github.com/SkygearIO/features/issues/415